code/modules/power/supermatter/supermatter.dm

```diff
@@ -197,151 +206,130 @@
 
 	//Ok, get the air from the turf
 	var/datum/gas_mixture/env = T.return_air()
-
 	var/datum/gas_mixture/removed
-
 	if(produces_gas)
 		//Remove gas from surrounding area
 		removed = env.remove(gasefficency * env.total_moles())
 	else
 		// Pass all the gas related code an empty gas container
 		removed = new()
 
-	if(!removed || !removed.total_moles() || isspaceturf(T)) //we're in space or there is no gas to process
-		if(takes_damage)
-			damage += max((power-1600)/10, 0)
-		return 1
+	//Check if we have gas.
+	if(!removed || !removed.total_moles() || isspaceturf(T))
+		has_gas = FALSE
+	else
+		has_gas = TRUE
 
-	damage_archived = damage
+	//Process negative effects.
+	if(power > POWER_PENALTY_THRESHOLD || damage > damage_penalty_point)
+		SM_autoprocess_lightning()
+		SM_autoprocess_anomaly()
+
+	//Process powerloss
+	SM_autoprocess_powerloss()
+
+	//Process radiation
+	SM_autoprocess_proximity()
+
+
+	//Process damage.
 	if(takes_damage)
-		//causing damage
-		damage = max(damage + (max(removed.temperature - ((T0C + HEAT_PENALTY_THRESHOLD)*dynamic_heat_resistance), 0) * mole_heat_penalty / 150 ) * DAMAGE_INCREASE_MULTIPLIER, 0)
-		damage = max(damage + (max(power - POWER_PENALTY_THRESHOLD, 0)/500) * DAMAGE_INCREASE_MULTIPLIER, 0)
-		damage = max(damage + (max(combined_gas - MOLE_PENALTY_THRESHOLD, 0)/80) * DAMAGE_INCREASE_MULTIPLIER, 0)
+		SM_autoprocess_damage(removed, T)
 
-		//healing damage
-		if(combined_gas < MOLE_PENALTY_THRESHOLD)
-			damage = max(damage + (min(removed.temperature - (T0C + HEAT_PENALTY_THRESHOLD), 0) / 150 ), 0)
+	//Process matter
+	SM_autoprocess_matter()
 
-		//capping damage
-		damage = min(damage_archived + (DAMAGE_HARDCAP * explosion_point),damage)
-		if(damage > damage_archived && prob(10))
-			playsound(get_turf(src), 'sound/effects/empulse.ogg', 50, 1)
+	//Process alerts
+	SM_autoprocess_alert()
+
+	//Process explosion check
+	SM_autoprocess_explosioncheck()
+
+	//Process temperature factor
+	SM_autoprocess_tempfactor()
+
+	//Process gasses
+	SM_autoprocess_gas(removed, env)
 
+	return TRUE
+
+/obj/machinery/power/supermatter_shard/proc/SM_autoprocess_gas(datum/gas_mixture/removed, datum/gas_mixture/env)
 	removed.assert_gases("o2", "plasma", "co2", "n2o", "n2", "freon")
```

Comment:
I'm not sure. I feel like total_moles() might throw a runtime if the mixture is empty.
Well, if it works, then you still have to rewrite the 6 lines just below this one since a 0 mole mix would end up causing division by zero.